File: bilibili_anime_scrape/scrape_bilibili_anime.mjs
// scrape_bilibili_anime.js
import puppeteer from "puppeteer";

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// Prefer the best cover URL from <picture>, else <img>, normalize protocol
function extractCoverURL(card) {
  const pic = card.querySelector(".season-cover-img picture");
  // Prefer AVIF/WebP source; fall back to <img>
  const pickFromSource = (type) => {
    const s = pic?.querySelector(`source[type="${type}"]`);
    if (!s) return null;
    const srcset = s.getAttribute("srcset") || "";
    const first = srcset.split(",")[0]?.trim().split(" ")[0];
    return first || null;
  };

  let url =
    pickFromSource("image/avif") ||
    pickFromSource("image/webp") ||
    pic?.querySelector("img")?.getAttribute("src") ||
    card.querySelector(".season-cover-img img")?.getAttribute("src") ||
    null;

  if (!url) return null;
  if (url.startsWith("//")) url = "https:" + url;
  if (url.startsWith("/")) url = "https://www.bilibili.com" + url;
  return url;
}

(async () => {
  // const browser = await puppeteer.launch({
  //   headless: true, // or "new"
  //   args: [
  //     "--no-sandbox",
  //     "--disable-setuid-sandbox",
  //     "--disable-dev-shm-usage",
  //   ],
  //   defaultViewport: { width: 1280, height: 1200 },
  // });

  const browser = await puppeteer.launch({
    executablePath: "/usr/bin/chromium", // or /usr/bin/google-chrome-stable
    headless: true,
  });
  const page = await browser.newPage();
  // Mildly realistic UA; reduces anti-bot friction
  await page.setUserAgent(
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) " +
      "Chrome/120.0.0.0 Safari/537.36",
  );

  // Basic request hygiene: block heavy assets we don’t need
  await page.setRequestInterception(true);
  page.on("request", (req) => {
    const type = req.resourceType();
    if (["stylesheet", "font"].includes(type)) return req.abort();
    return req.continue();
  });

  await page.goto("https://www.bilibili.com/anime/", {
    waitUntil: "domcontentloaded",
    timeout: 60_000,
  });

  // Helper to detect network idleness between scroll chunks
  async function waitForNetworkQuiet(ms = 1000, timeout = 15_000) {
    let idleResolve;
    let inflight = 0;
    let idleTimer;

    function resetIdle() {
      clearTimeout(idleTimer);
      idleTimer = setTimeout(() => idleResolve(), ms);
    }

    const rq = (req) => {
      inflight++;
      resetIdle();
    };
    const rs = (res) => {
      inflight = Math.max(0, inflight - 1);
      resetIdle();
    };
    const rf = (err) => {
      inflight = Math.max(0, inflight - 1);
      resetIdle();
    };

    const p = new Promise((resolve, reject) => {
      idleResolve = resolve;
      setTimeout(() => resolve(), timeout); // hard cap
    });

    page.on("request", rq);
    page.on("requestfinished", rs);
    page.on("requestfailed", rf);

    // start the idle window now
    resetIdle();
    await p;

    page.off("request", rq);
    page.off("requestfinished", rs);
    page.off("requestfailed", rf);
  }

  // Scroll until no new cards for a few iterations or we hit a sane cap
  let lastCount = 0;
  let stableRounds = 0;
  const MAX_ROUNDS = 20; // safety
  const STABLE_TARGET = 2; // stop when count stops growing twice in a row
  for (let round = 0; round < MAX_ROUNDS; round++) {
    await page.evaluate(() => window.scrollBy(0, window.innerHeight * 1.2));
    await sleep(400 + Math.random() * 300); // polite throttle
    await waitForNetworkQuiet(800, 8000);

    const count = await page.$$eval(
      ".hot-ranking-cell-wrapper",
      (els) => els.length,
    );
    if (count === lastCount) stableRounds++;
    else stableRounds = 0;

    lastCount = count;
    if (stableRounds >= STABLE_TARGET) break;
  }

  // Extract data from each card
  const data = await page.$$eval(".hot-ranking-cell-wrapper", (cards) => {
    const norm = (s) => (s || "").replace(/\s+/g, " ").trim();

    function extractCoverURL(card) {
      const pic = card.querySelector(".season-cover-img picture");
      const pickFromSource = (type) => {
        const s = pic?.querySelector(`source[type="${type}"]`);
        if (!s) return null;
        const srcset = s.getAttribute("srcset") || "";
        const first = srcset.split(",")[0]?.trim().split(" ")[0];
        return first || null;
      };
      let url =
        pickFromSource("image/avif") ||
        pickFromSource("image/webp") ||
        pic?.querySelector("img")?.getAttribute("src") ||
        card.querySelector(".season-cover-img img")?.getAttribute("src") ||
        null;

      if (!url) return null;
      if (url.startsWith("//")) url = "https:" + url;
      if (url.startsWith("/")) url = "https://www.bilibili.com" + url;
      return url;
    }

    return cards
      .map((card, idx) => {
        const cover = extractCoverURL(card);
        const titleEl = card.querySelector(".home-cell-desc-title");
        const subtitleEl = card.querySelector(".home-cell-desc-subtitle");
        const scoreEl = card.querySelector(".season-cover-score");
        const linkEl =
          card.querySelector("a.season-cover[href]") ||
          card.querySelector(".ranking-cell-desc a[href]");

        return {
          rank: idx + 1,
          title: norm(titleEl?.textContent),
          subtitle: norm(subtitleEl?.textContent),
          rating: norm(scoreEl?.textContent) || null, // e.g., "9.7" or null
          cover,
          href: linkEl
            ? linkEl.getAttribute("href")?.startsWith("//")
              ? "https:" + linkEl.getAttribute("href")
              : linkEl.getAttribute("href")
            : null,
        };
      })
      .filter((x) => x.title || x.cover);
  });

  console.log(JSON.stringify(data, null, 2));
  await browser.close();
})();

File: bilibili_anime_scrape/scrape_bilibili_anime.py
# scrape_bilibili_anime.py
import asyncio, json, re
from playwright.async_api import async_playwright

def normalize_url(u: str | None) -> str | None:
    if not u: return None
    if u.startswith("//"): return "https:" + u
    if u.startswith("/"):  return "https://www.bilibili.com" + u
    return u

async def main():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True, args=["--disable-dev-shm-usage"])
        ctx = await browser.new_context(
            viewport={"width": 1280, "height": 1200},
            user_agent=("Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
                        "(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"),
        )
        page = await ctx.new_page()
        # Reduce bandwidth: block fonts/css
        await ctx.route("**/*", lambda route: route.abort() if route.request.resource_type in ("font","stylesheet") else route.continue_())
        await page.goto("https://www.bilibili.com/anime/", wait_until="domcontentloaded", timeout=60000)

        async def network_quiet(ms=800, timeout=8000):
            # crude quiet wait
            await page.wait_for_timeout(ms)

        # scroll loop
        last = 0; stable = 0
        for _ in range(20):
            await page.mouse.wheel(0, 1000)
            await network_quiet()
            count = await page.locator(".hot-ranking-cell-wrapper").count()
            if count == last: stable += 1
            else: stable = 0
            last = count
            if stable >= 2: break

        cards = page.locator(".hot-ranking-cell-wrapper")
        n = await cards.count()
        out = []
        for i in range(n):
            card = cards.nth(i)
            # cover via picture > source (avif/webp) fallback to img
            url = await card.locator(".season-cover-img picture source[type='image/avif']").first.get_attribute("srcset")
            if not url:
                url = await card.locator(".season-cover-img picture source[type='image/webp']").first.get_attribute("srcset")
            if url:
                url = url.split(",")[0].strip().split(" ")[0]
            if not url:
                url = await card.locator(".season-cover-img img").first.get_attribute("src")
            url = normalize_url(url)

            title = (await card.locator(".home-cell-desc-title").inner_text()).strip() if await card.locator(".home-cell-desc-title").count() else None
            subtitle = (await card.locator(".home-cell-desc-subtitle").inner_text()).strip() if await card.locator(".home-cell-desc-subtitle").count() else None
            rating = (await card.locator(".season-cover-score").inner_text()).strip() if await card.locator(".season-cover-score").count() else None

            href = await card.locator("a.season-cover").first.get_attribute("href")
            href = normalize_url(href)

            out.append({
                "rank": i+1, "title": title, "subtitle": subtitle,
                "rating": rating, "cover": url, "href": href
            })

        print(json.dumps([x for x in out if x["title"] or x["cover"]], ensure_ascii=False, indent=2))
        await browser.close()

if __name__ == "__main__":
    asyncio.run(main())
